streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java

```diff
@@ -74,20 +209,240 @@ public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareStandbyTas
+     * If a logical task has no source topic partitions, it will return false.
+     * If standby tasks are configured, but a logical task has no changelog topic partitions, it will return false.
      *
-     * @return a new map containing an assignment that replicates exactly the previous assignment reported
-     *         in the applicationState
+     * @return whether rack-aware assignment decisions can be made for this application.
      */
-    public static Map<ProcessId, KafkaStreamsAssignment> identityAssignment(
-        final ApplicationState applicationState
-    ) {
-        throw new UnsupportedOperationException("Not Implemented.");
+    public static boolean hasValidRackInformation(final ApplicationState applicationState) {
+        for (final KafkaStreamsState state : applicationState.kafkaStreamsStates(false).values()) {
+            if (!hasValidRackInformation(state)) {
+                return false;
+            }
+        }
+
+        final boolean hasStandby = applicationState.assignmentConfigs().numStandbyReplicas() >= 1;
+        for (final TaskInfo task : applicationState.allTasks()) {
+            if (!hasValidRackInformation(task, hasStandby)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private static boolean hasValidRackInformation(final KafkaStreamsState state) {
+        if (!state.rackId().isPresent()) {
+            LOG.error("Client " + state.processId() + " doesn't have rack configured.");
+            return false;
+        }
+        return true;
+    }
+
+    private static boolean hasValidRackInformation(final TaskInfo task, final boolean hasStandby) {
+        final Map<TopicPartition, Set<String>> racksByTopicPartition = task.partitionToRackIds();
+        for (final TopicPartition topicPartition : task.sourceTopicPartitions()) {
+            final Set<String> racks = racksByTopicPartition.getOrDefault(topicPartition, null);
+            if (racks == null || racks.isEmpty()) {
+                LOG.error("Topic partition {} for task {} does not have racks configured.", topicPartition, task.id());
+                return false;
+            }
+        }
+
+        for (final TopicPartition topicPartition : task.changelogTopicPartitions()) {
+            final Set<String> racks = racksByTopicPartition.getOrDefault(topicPartition, null);
+            if (racks == null || racks.isEmpty()) {
+                LOG.error("Topic partition {} for task {} does not have racks configured.", topicPartition, task.id());
+                return false;
+            }
+        }
+
+        if (task.sourceTopicPartitions().isEmpty()) {
+            LOG.error("Task {} has no source TopicPartitions.", task.id());
+            return false;
+        }
+
+        if (hasStandby && task.changelogTopicPartitions().isEmpty()) {
+            LOG.error("Task {} has no changelog TopicPartitions.", task.id());
+            return false;
```

Comment:
This should not be an error, we can remove the `hasStandby` parameter entirely.

For one thing, it shouldn't be an error to have configured an app to use standby tasks even if there's no state/changelog topics (Streams just wouldn't create those standby tasks). But for another, it's also possible to have a standby task with no changelog topics. Not all stateful tasks necessarily have to have changelogs (though most do), one can configure a state store with "logging disabled" which just means it has no changelog topic. Sometimes people play fast and lose with semantics